### ISSUE TITLE
Malloc tutorial

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#5009]
Malloc tutorial<br>
<br>
https://ift.tt/2pC5Xrm<br>
<br>
<br>
<br>
<div>
<strong>Malloc tutorial</strong>
<hr>
<p>Let's write a <a href="http://man7.org/linux/man-pages/man3/malloc.3.html">malloc</a> and see how it works with existing programs!</p>
<p>This tutorial is going to assume that you know what pointers are, and that you know enough C to know that <code>*ptr</code> dereferences a pointer, <code>ptr-&gt;foo</code> means <code>(*ptr).foo</code>, that malloc is used to <a href="http://duartes.org/gustavo/blog/post/how-the-kernel-manages-your-memory/">dynamically allocate space</a>, and that you're familiar with the concept of a linked list. For a basic intro to C, <a href="http://www.amazon.com/gp/product/0673999866/ref=as_li_qf_sp_asin_il_tl?ie=UTF8&amp;camp=1789&amp;creative=9325&amp;creativeASIN=0673999866&amp;linkCode=as2&amp;tag=abroaview-20&amp;linkId=5C3DNUKAQELP2KUL">Pointers on C</a> is one of my favorite books. If you want to look at all of this code at once, it's available <a href="https://github.com/danluu/malloc-tutorial/blob/master/malloc.c">here</a>.</p>
<p>Preliminaries aside, malloc's function signature is</p>
<pre>
<code>void *malloc(size_t size);
</code>
</pre>
<p>It takes as input a number of bytes and returns a pointer to a block of memory of that size.</p>
<p>There are a number of ways we can implement this. We're going to arbitrarily choose to use the <a href="http://man7.org/linux/man-pages/man2/sbrk.2.html">sbrk</a> syscall. The OS reserves stack and heap space for processes and sbrk lets us manipulate the heap. <code>sbrk(0)</code> returns a pointer to the current top of the heap. <code>sbrk(foo)</code> increments the heap size by <code>foo</code> and returns a pointer to the previous top of the heap.</p>
<p><img alt="Diagram of linux memory layout, courtesy of Gustavo Duarte." width="194" src="https://danluu.com/malloc-tutorial/../images/malloc-tutorial/heap.png" height="301"></p>
<p>If we want to implement a really simple malloc, we can do something like</p>
<pre>
<code>#include &lt;assert.h&gt;
#include &lt;string.h&gt;
#include &lt;sys/types.h&gt;
#include &lt;unistd.h&gt;

void *malloc(size_t size) {
  void *p = sbrk(0);
  void *request = sbrk(size);
  if (request == (void*) -1) {
    return NULL; // sbrk failed.
  } else {
    assert(p == request); // Not thread safe.
    return p;
  }
}
</code>
</pre>
<p>When a program asks malloc for space, malloc asks sbrk to increment the heap size and returns a pointer to the start of the new region on the heap. This is missing a technicality, that <code>malloc(0)</code> should either return <code>NULL</code> or another pointer that can be passed to free without causing havoc, but it basically works.</p>
<p>But speaking of free, how does free work? Free's prototype is</p>
<pre>
<code>void free(void *ptr);
</code>
</pre>
<p>When free is passed a pointer that was previously returned from malloc, it's supposed to free the space. But given a pointer to something allocated by our malloc, we have no idea what size block is associated with it. Where do we store that? If we had a working malloc, we could malloc some space and store it there, but we're going to run into trouble if we need to call malloc to reserve space each time we call malloc to reserve space.</p>
<p>A common trick to work around this is to store meta-information about a memory region in some space that we squirrel away just below the pointer that we return. Say the top of the heap is currently at <code>0x1000</code> and we ask for <code>0x400</code> bytes. Our current malloc will request <code>0x400</code> bytes from <code>sbrk</code> and return a pointer to <code>0x1000</code>. If we instead save, say, <code>0x10</code> bytes to store information about the block, our malloc would request <code>0x410</code> bytes from <code>sbrk</code> and return a pointer to <code>0x1010</code>, hiding our <code>0x10</code> byte block of meta-information from the code that's calling malloc.</p>
<p>That lets us free a block, but then what? The heap region we get from the OS has to be contiguous, so we can't return a block of memory in the middle to the OS. Even if we were willing to copy everything above the newly freed region down to fill the hole, so we could return space at the end, there's no way to notify all of the code with pointers to the heap that those pointers need to be adjusted.</p>
<p>Instead, we can mark that the block has been freed without returning it to the OS, so that future calls to malloc can use re-use the block. But to do that we'll need be able to access the meta information for each block. There are a lot of possible solutions to that. We'll arbitrarily choose to use a single linked list for simplicity.</p>
<p>So, for each block, we'll want to have something like</p>
<pre>
<code>struct block_meta {
  size_t size;
  struct block_meta *next;
  int free;
  int magic; // For debugging only. TODO: remove this in non-debug mode.
};

#define META_SIZE sizeof(struct block_meta)
</code>
</pre>
<p>We need to know the size of the block, whether or not it's free, and what the next block is. There's a magic number here for debugging purposes, but it's not really necessary; we'll set it to arbitrary values, which will let us easily see which code modified the struct last.</p>
<p>We'll also need a head for our linked list:</p>
<pre>
<code>void *global_base = NULL;
</code>
</pre>
<p>For our malloc, we'll want to re-use free space if possible, allocating space when we can't re-use existing space. Given that we have this linked list structure, checking if we have a free block and returning it is straightforward. When we get a request of some size, we iterate through our linked list to see if there's a free block that's large enough.</p>
<pre>
<code>struct block_meta *find_free_block(struct block_meta **last, size_t size) {
  struct block_meta *current = global_base;
  while (current &amp;&amp; !(current-&gt;free &amp;&amp; current-&gt;size &gt;= size)) {
    *last = current;
    current = current-&gt;next;
  }
  return current;
}
</code>
</pre>
<p>If we don't find a free block, we'll have to request space from the OS using sbrk and add our new block to the end of the linked list.</p>
<pre>
<code>struct block_meta *request_space(struct block_meta* last, size_t size) {
  struct block_meta *block;
  block = sbrk(0);
  void *request = sbrk(size + META_SIZE);
  assert((void*)block == request); // Not thread safe.
  if (request == (void*) -1) {
    return NULL; // sbrk failed.
  }

  if (last) { // NULL on first request.
    last-&gt;next = block;
  }
  block-&gt;size = size;
  block-&gt;next = NULL;
  block-&gt;free = 0;
  block-&gt;magic = 0x12345678;
  return block;
}
</code>
</pre>
<p>As with our original implementation, we request space using <code>sbrk</code>. But we add a bit of extra space to store our struct, and then set the fields of the struct appropriately.</p>
<p>Now that we have helper functions to check if we have existing free space and to request space, our malloc is simple. If our global base pointer is <code>NULL</code>, we need to request space and set the base pointer to our new block. If it's not <code>NULL</code>, we check to see if we can re-use any existing space. If we can, then we do; if we can't, then we request space and use the new space.</p>
<pre>
<code>void *malloc(size_t size) {
  struct block_meta *block;
  // TODO: align size?

  if (size &lt;= 0) {
    return NULL;
  }

  if (!global_base) { // First call.
    block = request_space(NULL, size);
    if (!block) {
      return NULL;
    }
    global_base = block;
  } else {
    struct block_meta *last = global_base;
    block = find_free_block(&amp;last, size);
    if (!block) { // Failed to find free block.
      block = request_space(last, size);
      if (!block) {
        return NULL;
      }
    } else {      // Found free block
      // TODO: consider splitting block here.
      block-&gt;free = 0;
      block-&gt;magic = 0x77777777;
    }
  }

  return(block+1);
}
</code>
</pre>
<p>For anyone who isn't familiar with C, we return block+1 because we want to return a pointer to the region after block_meta. Since block is a pointer of type <code>struct block_meta</code>, <code>+1</code> increments the address by one <code>sizeof(struct(block_meta))</code>.</p>
<p>If we just wanted a malloc without a free, we could have used our original, much simpler malloc. So let's write free! The main thing free needs to do is set <code>-&gt;free</code>.</p>
<p>Because we'll need to get the address of our struct in multiple places in our code, let's define this function.</p>
<pre>
<code>struct block_meta *get_block_ptr(void *ptr) {
  return (struct block_meta*)ptr - 1;
}
</code>
</pre>
<p>Now that we have that, here's free:</p>
<pre>
<code>void free(void *ptr) {
  if (!ptr) {
    return;
  }

  // TODO: consider merging blocks once splitting blocks is implemented.
  struct block_meta* block_ptr = get_block_ptr(ptr);
  assert(block_ptr-&gt;free == 0);
  assert(block_ptr-&gt;magic == 0x77777777 || block_ptr-&gt;magic == 0x12345678);
  block_ptr-&gt;free = 1;
  block_ptr-&gt;magic = 0x55555555;
}
</code>
</pre>
<p>In addition to setting <code>-&gt;free</code>, it's valid to call free with a NULL ptr, so we need to check for NULL. Since free shouldn't be called on arbitrary addresses or on blocks that are already freed, we can assert that those things never happen.</p>
<p>You never really need to assert anything, but it often makes debugging a lot easier. In fact, when I wrote this code, I had a bug that would have resulted in silent data corruption if these asserts weren't there. Instead, the code failed at the assert, which make it trivial to debug.</p>
<p>Now that we've got malloc and free, we can write programs using our custom memory allocator! But before we can drop our allocator into existing code, we'll need to implement a couple more common functions, realloc and calloc. Calloc is just malloc that initializes the memory to 0, so let's look at realloc first. Realloc is supposed to adjust the size of a block of memory that we've gotten from malloc, calloc, or realloc.</p>
<p>Realloc's function prototype is</p>
<pre>
<code>void *realloc(void *ptr, size_t size)
</code>
</pre>
<p>If we pass realloc a NULL pointer, it's supposed to act just like malloc. If we pass it a previously malloced pointer, it should free up space if the size is smaller than the previous size, and allocate more space and copy the existing data over if the size is larger than the previous size.</p>
<p>Everything will still work if we don't resize when the size is decreased and we don't free anything, but we absolutely have to allocate more space if the size is increased, so let's start with that.</p>
<pre>
<code>void *realloc(void *ptr, size_t size) {
  if (!ptr) {
    // NULL ptr. realloc should act like malloc.
    return malloc(size);
  }

  struct block_meta* block_ptr = get_block_ptr(ptr);
  if (block_ptr-&gt;size &gt;= size) {
    // We have enough space. Could free some once we implement split.
    return ptr;
  }

  // Need to really realloc. Malloc new space and free old space.
  // Then copy old data to new space.
  void *new_ptr;
  new_ptr = malloc(size);
  if (!new_ptr) {
    return NULL; // TODO: set errno on failure.
  }
  memcpy(new_ptr, ptr, block_ptr-&gt;size);
  free(ptr);
  return new_ptr;
}
</code>
</pre>
<p>And now for calloc, which just clears the memory before returning a pointer.</p>
<pre>
<code>void *calloc(size_t nelem, size_t elsize) {
  size_t size = nelem * elsize; // TODO: check for overflow.
  void *ptr = malloc(size);
  memset(ptr, 0, size);
  return ptr;
}
</code>
</pre>
<p>Note that this doesn't check for overflow in <code>nelem * elsize</code>, which is actually required by the spec. All of the code here is just enough to get something that kinda sorta works.</p>
<p>Now that we have something that kinda works, we can use our with existing programs (and we don't even need to recompile the programs)!</p>
<p>First, we need to compile our code. On linux, something like</p>
<pre>
<code>clang -O0 -g -W -Wall -Wextra -shared -fPIC malloc.c -o malloc.so
</code>
</pre>
<p>should work.</p>
<p><code>-g</code> adds debug symbols, so we can look at our code with <code>gdb</code> or <code>lldb</code>. <code>-O0</code> will help with debugging, by preventing individual variables from getting optimized out. <code>-W -Wall -Wextra</code> adds extra warnings. <code>-shared -fPIC</code> will let us dynamically link our code, which is what lets us <a href="http://jvns.ca/blog/2014/11/27/ld-preload-is-super-fun-and-easy/">use our code with existing binaries</a>!</p>
<p>On macs, we'll want something like</p>
<pre>
<code>clang -O0 -g -W -Wall -Wextra -dynamiclib malloc.c -o malloc.dylib
</code>
</pre>
<p>Note that <code>sbrk</code> is deprecated on recent versions of OS X. Apple uses an unorthodox definition of deprecated -- some deprecated syscalls are badly broken. I didn't really test this on a Mac, so it's possible that this will cause weird failures or or just not work on a mac.</p>
<p>Now, to use get a binary to use our malloc on linux, we'll need to set the <code>LD_PRELOAD</code> environment variable. If you're using bash, you can do that with</p>
<pre>
<code>export LD_PRELOAD=/absolute/path/here/malloc.so
</code>
</pre>
<p>If you've got a mac, you'll want</p>
<pre>
<code>export DYLD_INSERT_LIBRARIES=/absolute/path/here/malloc.so
</code>
</pre>
<p>If everything works, you can run some arbitrary binary and it will run as normal (except that it will be a bit slower).</p>
<pre>
<code>$ ls
Makefile  malloc.c  malloc.so  README.md  test  test-0  test-1  test-2  test-3  test-4
</code>
</pre>
<p>If there's a bug, you might get something like</p>
<pre>
<code>$ ls
Segmentation fault (core dumped)
</code>
</pre>
<h4>Debugging</h4>
<p>Let's talk about debugging! If you're familiar with using a debugger to set breakpoints, inspect memory, and step through code, you can skip this section and go straight to <a href="#exercises">the exercises</a>.</p>
<p>This section assumes you can figure out how to install gdb on your system. If you're on a mac, you may want to just use lldb and translate the commands appropriately. Since I don't know what bugs you might run into, I'm going to introduce a couple of bugs and show how I'd track them down.</p>
<p>First, we need to figure out how to run gdb without having it segfault. If ls segfaults, and we try to run <code>gdb ls</code>, gdb is almost certainly going to segfault, too. We could write a wrapper to do this, but gdb also supports this. If we start gdb and then run <code>set environment LD_PRELOAD=./malloc.so</code> before running the program, <code>LD_PRELOAD</code> will work as normal.</p>
<pre>
<code>$ gdb /bin/ls
(gdb) set environment LD_PRELOAD=./malloc.so
(gdb) run
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7bd7dbd in free (ptr=0x0) at malloc.c:113
113       assert(block_ptr-&gt;free == 0);
</code>
</pre>
<p>As expected, we get a segfault. We can look around with <code>list</code> to see the code near the segfault.</p>
<pre>
<code>(gdb) list
108     }
109
110     void free(void *ptr) {
111       // TODO: consider merging blocks once splitting blocks is implemented.
112       struct block_meta* block_ptr = get_block_ptr(ptr);
113       assert(block_ptr-&gt;free == 0);
114       assert(block_ptr-&gt;magic == 0x77777777 || block_ptr-&gt;magic == 0x12345678);
115       block_ptr-&gt;free = 1;
116       block_ptr-&gt;magic = 0x55555555;
117     }
</code>
</pre>
<p>And then we can use <code>p</code> (for print) to see what's going on with the variables here:</p>
<pre>
<code>(gdb) p ptr
$6 = (void *) 0x0
(gdb) p block_ptr
$7 = (struct block_meta *) 0xffffffffffffffe8
</code>
</pre>
<p><code>ptr</code> is <code>0</code>, i.e., <code>NULL</code>, which is the cause of the problem: we forgot to check for NULL.</p>
<p>Now that we've figured that out, let's try a slightly harder bug. Let's say that we decided to replace our struct with</p>
<pre>
<code>struct block_meta {
  size_t size;
  struct block_meta *next;
  int free;
  int magic;    // For debugging only. TODO: remove this in non-debug mode.
  char data[1];
};
</code>
</pre>
<p>and then return <code>block-&gt;data</code> instead of <code>block+1</code> from malloc, with no other changes. This seems pretty similar to what we're already doing -- we just define a member that points to the end of the struct, and return a pointer to that.</p>
<p>But here's what happens if we try to use our new malloc:</p>
<pre>
<code>$ /bin/ls
Segmentation fault (core dumped)
gdb /bin/ls
(gdb) set environment LD_PRELOAD=./malloc.so
(gdb) run

Program received signal SIGSEGV, Segmentation fault.
_IO_vfprintf_internal (s=s@entry=0x7fffff7ff5f0, format=format@entry=0x7ffff7567370 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", ap=ap@entry=0x7fffff7ff718) at vfprintf.c:1332
1332    vfprintf.c: No such file or directory.
1327    in vfprintf.c
</code>
</pre>
<p>This isn't as nice as our last error -- we can see that one of our asserts failed, but gdb drops us into some print function that's being called when the assert fails. But that print function uses our buggy malloc and blows up!</p>
<p>One thing we could do from here would be to inspect <code>ap</code> to see what <code>assert</code> was trying to print:</p>
<pre>
<code>(gdb) p *ap
$4 = {gp_offset = 16, fp_offset = 48, overflow_arg_area = 0x7fffff7ff7f0, reg_save_area = 0x7fffff7ff730}
</code>
</pre>
<p>That would work fine; we could poke around until we figure out what's supposed to get printed and figure out the fail that way. Some other solutions would be to write our own custom assert or to use the right hooks to prevent <code>assert</code> from using our malloc.</p>
<p>But in this case, we know there are only a few asserts in our code. The one in malloc checking that we don't try to use this in a multithreaded program and the two in free checking that we're not freeing something we shouldn't. Let's look at free first, by setting a breakpoint.</p>
<pre>
<code>$ gdb /bin/ls
(gdb) set environment LD_PRELOAD=./malloc.so
(gdb) break free
Breakpoint 1 at 0x400530
(gdb) run /bin/ls

Breakpoint 1, free (ptr=0x61c270) at malloc.c:112
112       if (!ptr) {
</code>
</pre>
<p><code>block_ptr</code> isn't set yet, but if we use <code>s</code> a few times to step forward to after it's set, we can see what the value is:</p>
<pre>
<code>(gdb) s
(gdb) s
(gdb) s
free (ptr=0x61c270) at malloc.c:118
118       assert(block_ptr-&gt;free == 0);
(gdb) p/x *block_ptr
$11 = {size = 0, next = 0x78, free = 0, magic = 0, data = ""}
</code>
</pre>
<p>I'm using <code>p/x</code> instead of <code>p</code> so we can see it in hex. The <code>magic</code> field is 0, which should be impossible for a valid struct that we're trying to free. Maybe <code>get_block_ptr</code> is returning a bad offset? We have <code>ptr</code> available to us, so we can just inspect different offsets. Since it's a <code>void *</code>, we'll have to cast it so that gdb knows how to interpret the results.</p>
<pre>
<code>(gdb) p sizeof(struct block_meta)
$12 = 32
(gdb) p/x *(struct block_meta*)(ptr-32)
$13 = {size = 0x0, next = 0x78, free = 0x0, magic = 0x0, data = {0x0}}
(gdb) p/x *(struct block_meta*)(ptr-28)
$14 = {size = 0x7800000000, next = 0x0, free = 0x0, magic = 0x0, data = {0x78}}
(gdb) p/x *(struct block_meta*)(ptr-24)
$15 = {size = 0x78, next = 0x0, free = 0x0, magic = 0x12345678, data = {0x6e}}
</code>
</pre>
<p>If we back off a bit from the address we're using, we can see that the correct offset is 24 and not 32. What's happening here is that structs get padded, so that <code>sizeof(struct block_meta)</code> is 32, even though the last valid member is at <code>24</code>. If we want to cut out that extra space, we need to fix <code>get_block_ptr</code>.</p>
<p>That's it for debugging!</p>
<h4>Exercises</h4>
<p>Personally, this sort of thing never sticks with me unless I work through some exercsies, so I'll leave a couple exercises here for anyone who's interested.</p>
<ol>
<li>
<p>malloc is supposed to return a pointer &ldquo;which is suitably aligned for any built-in type&rdquo;. Does our malloc do that? If so, why? If not, fix the alignment. Note that &ldquo;any built-in type&rdquo; is basically up to 8 bytes for C because SSE/AVX types aren't built-in types.</p>
</li>
<li>
<p>Our malloc is really wasteful if we try to re-use an existing block and we don't need all of the space. Implement a function that will split up blocks so that they use the minimum amount of space necessary</p>
</li>
<li>
<p>After doing <code>2</code>, if we call malloc and free lots of times with random sizes, we'll end up with a bunch of small blocks that can only be re-used when we ask for small amounts of space. Implement a mechanism to merge adjacent free blocks together so that any consecutive free blocks will get merged into a single block.</p>
</li>
<li>
<p>Find bugs in the existing code! I haven't tested this much, so I'm sure there are bugs, even if this basically kinda sorta works.</p>
</li>
</ol>
<h4>Resources</h4>
<p>I read <a href="http://www.inf.udec.cl/~leo/Malloc_tutorial.pdf">this tutorial</a> by Marwan Burelle before sitting down and trying to write my own implementation, so it's pretty similar. The main implementaiton differences are that my version is simpler, but more vulnerable to memory fragmentation. In terms of exposition, my style is a lot more casual. If you want something a bit more formal, Dr. Burelle has you covered.</p>
<p>For more on how Linux deals with memory management, see <a href="http://duartes.org/gustavo/blog/post/anatomy-of-a-program-in-memory/">this post</a> by Gustavo Duarte.</p>
<p>For more on how real-world malloc implementations work, <a href="http://g.oswego.edu/dl/html/malloc.html">dlmalloc</a> and <a href="http://goog-perftools.sourceforge.net/doc/tcmalloc.html">tcmalloc</a> are both great reading. I haven't read the code for <a href="http://www.canonware.com/jemalloc/">jemalloc</a>, and I've heard that it's a bit more more difficult to understand, but it's also the most widely used high-performance malloc implementation around.</p>
<p>For help debugging, <a href="https://code.google.com/p/address-sanitizer/wiki/AddressSanitizer">Address Sanitizer</a> is amazing. If you want to write a thread-safe version, <a href="https://code.google.com/p/data-race-test/wiki/ThreadSanitizer">Thread Sanitizer</a> is also a great tool.</p>
<p>There's a <a href="http://mgarciaisaia.github.io/tutorial-c/blog/2014/12/26/un-tutorial-rapido-para-implementar-y-debuggear-malloc/">spanish translation of this post here</a> thanks to Matias Garcia Isaia.</p>
<h4>Acknowledgements</h4>
<p>Thanks to Gustavo Duarte for letting me use one of his images to illustrate sbrk, to Ian Whitlock and Danielle Sucher for finding some typos, to Nathan Kurz for suggestions on additional resources, and to "tedu" for catching a bug. Please <a href="https://twitter.com/danluu">let me know</a> if you find other bugs in this post (whether they're in the writing or the code).</p>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via danluu.com https://danluu.com<br>
<br>
August 2, 2019 at 11:58AM
